drivers: add new audio dai driver interface

juimonen opened the pull request:
The DAI (digital audio interface) API provides support for the standard
I2S (SSP) and its common variants. It supports also DMIC, HDA and SDW
backends. The API has a config function with bespoke data argument for
device/vendor specific config. There are also optional timestamping
functions to get device specific audio clock time.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

Commit:
drivers: add new audio dai driver interface

The DAI (digital audio interface) API is a high level audio driver
abstraction. It provides support for the standard I2S (SSP), DMIC, HDA
and SDW backends. The API has a config function with bespoke data
argument for device/vendor specific config. There are also optional
timestamping functions to get device specific audio clock time.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

teburd reviewed (commented):
I think this is, generally, a really nice API. I don't think init is needed here, and I wonder if a zero copy interface could be used instead of the copy-focused API presented here. I'm not the best person to judge on the copy front though as I haven't worked with codecs/dais/etc much.

arvid-r commented:
Is this meant to replace the I2S API at some point?

juimonen commented on include/drivers/dai.h:
```diff
@@ -0,0 +1,484 @@
+	DAI_TRIGGER_RESET,
+};
+
+/** \brief Properties of DAI
```
ok

juimonen commented:
@arvid-r As I see it, this API is supposed to be parallel or one abstraction level above I2S api. So you could use this API to do the same things as with I2S API. I had initially the I2S format definitions included, but most folks wanted to remove those as they kind of confuse the issue and make the file much harder to read. But probably those should be included somehow.

juimonen commented on include/drivers/dai.h:
```diff
@@ -0,0 +1,484 @@
+ * @endcond
+ */
+
+static inline int dai_init(const struct device *dev)
```
Existing drivers in sof (sound open firmware, which we are trying to port with this api), do dynamic memory allocation for parameters, do power-on for the hw, set up clocks, reset registers etc. with their "init" functions. As I understand the zephyr device init is more for device boot up (?). So I guess I would like to have more dynamic behavior like: "I'm going to soon start using SSP, so make everything ready for it", and when I'm done playback I can free the resources for others to use.

juimonen replied:
removed now, can be replaced with pm functions

juimonen commented on include/drivers/dai.h:
```diff
@@ -0,0 +1,484 @@
+ */
+static inline int dai_config_set(const struct device *dev,
+				 const struct dai_config *cfg,
+				 const void *bespoke_cfg)
```
yes agreed. So the idea is that this would be more general than I2S -> we would have different bespoke parameters for different drivers. Now for example our existing dmic driver has huge parameter set and it can operate in 2 different modes: typed nested structs or binary blobs with direct register layout. So when having several different drivers like this, I see only option to have a union of these structs as parameter and it is not looking very nice...

So I'm wondering how is the validation specified to be done? Is there any examples that one could send like non-typed data through the API and the driver implementation would be responsible of the validation (because the driver type understands what kind of data to expect)?

teburd replied:
void * unfortunately means, even if you have a pointer to some device specific config like struct my_dai_cfg{}, and a size of that struct pass as a param here so you could copy it, you can't guarantee the struct and the driver don't expect more pointers inside that cfg, which is a no-no from reading the docs. You need to be able to verify and copy any parameters that are used for branching based on the docs here:

https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-function

And that makes sense, if you have branch statements in supervisor mode based on data pass in, you need to ensure the data is 1) valid at the time of the call, 2) doesn't change once the call is started. Neither is possible really with a void*.

Really though this *only*  matters if you care about ever supporting usermode syscalls to this particular API. You could simply say this is a supervisor only call, and not sure the __syscall macro. That doesn't prevent using __syscalls with other APIs here either, everything else seems perfectly valid.

teburd replied:
What the gist of the comment above boils down to, is void* won't preclude the API from being a syscall at some point but that there might need to be some inventive thinking around it. That's ok, marking this as resolved.

arvid-r commented:
The biggest problem I had with the I2S API is the fact that you have to initialise the RX part and TX part of a codec with separate init calls to the same device, which is very confusing. But here we have `DAI_DIR_BOTH` which is great. It is probably good to clarify if that has to be supported, or if you could to separate calls to DAI_DIR_RX and DAI_DIR_TX

juimonen commented:
@lgirdwood updated the PR according to comments from API call:
- removed init/deinit as their intended functionality corresponds to zephyr device pm functions
- removed read/write as they are copies from legacy (adapted from video api) and sof is not using them (we use dma transfers)

mnkp commented:
@juimonen Could you add to the PR description some names of actual hardware devices / modules that the new DAI driver will be used for. And if you also included links to datasheets would be even better.

juimonen commented:
@mnkp you mean to the actual commit message or here in the PR as info? 

We are going to port all sof (sound open firmware) drivers to zephyr with this interface, so most of the hw will be audio devices embedded to intel soc's, like ssp, dmic, hda, alh/soundwire. There are some drivers from nxp also in sof code.

Even though I work for Intel I'm not totally sure how much the actual low level hw specs are public, or just available if public :)

You can take a look at sof code (https://github.com/thesofproject/sof), so it would be a quite wide range of audio devices available in different generations of intel socs. As said there are couple of other manufacturers involved and we've been using this kind of interface for wide range of devices, but I can't be 100% sure of the fit for everybody else. However, I have certain practical confidence as we've been using it for long time.

If there are extra needs we can revisit this API in the experimental phase.

carlescufi commented:
@anangl could you please take a look at the proposal?

anangl commented on include/drivers/dai.h:
```diff
@@ -0,0 +1,462 @@
+{
+	const struct dai_driver_api *api = (const struct dai_driver_api *)dev->api;
+
+	if (!api->ts_config)
+		return -EINVAL;
```
When a function is not implemented, `-ENOSYS` should be returned. See [API Design Guidelines](https://docs.zephyrproject.org/latest/reference/api/design_guidelines.html#return-codes).
Braces should also be added (see [Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style)).

mbolivar-nordic reviewed (changes requested):
Please, no new APIs without documentation entries... we keep doing this (https://github.com/zephyrproject-rtos/zephyr/issues/42842 and almost https://github.com/zephyrproject-rtos/zephyr/pull/41856 if I hadn't happened to look at it), now this PR too.

This needs an entry in doc/reference/api/overview.rst and a corresponding page in the docs.

nashif commented:
> Please, no new APIs without documentation entries...

missed that, completely agree

nashif commented:
@juimonen  please rebase.

nashif commented:
part of https://github.com/zephyrproject-rtos/zephyr/pull/43905
